rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,783 @@
+              returnedItem = childItemOpt match {
+                case None => none
+                case Some(childItem) =>
+                  val updatedChildNode = childNode.updated(childItemIdx, childItem)
+                  saveNodeAndCreateItem(updatedChildNode, commPrefix, optimization = false).some
+              }
+            } yield returnedItem
+          } else {
+            // Create child node, insert existing Ptr and new leaf in this node
+            val newNode = emptyNode
+              .updated(byteToInt(ptrPrefixRest.head), NodePtr(ptrPrefixRest.tail, ptr))
+              .updated(byteToInt(insPrefixRest.head), Leaf(insPrefixRest.tail, insValue))
+            saveNodeAndCreateItem(newNode, commPrefix, optimization = false).some.pure
+          }
+      }
+
+    /**
+      * Delete leaf value from this part of tree (start from curItem).
+      * Rehash and save all depend node to cacheW. Return updated curItem.
+      * If not found leaf with  delPrefix - return [[None]].
+      */
+    def delete(curItem: Item, delPrefix: ByteVector): F[Option[Item]] =
+      curItem match {
+        case EmptyItem => none[Item].pure //Not found
+
+        case Leaf(leafPrefix, _) =>
+          if (leafPrefix == delPrefix) (EmptyItem: Item).some.pure //Happy end
+          else none[Item].pure                                     //Not found
+
+        case NodePtr(ptrPrefix, ptr) =>
+          val (commPrefix, delPrefixRest, ptrPrefixRest) = commonPrefix(delPrefix, ptrPrefix)
+          if (ptrPrefixRest.nonEmpty || delPrefixRest.isEmpty) none[Item].pure //Not found
+          else {
+            val (childItemIdx, childDelPrefix) =
+              (byteToInt(delPrefixRest.head), delPrefixRest.tail)
+            for {
+              childNode    <- loadNode(ptr)
+              childItemOpt <- delete(childNode(childItemIdx), childDelPrefix) //deeper
+              returnedChild = childItemOpt match {
+                case None => none
+                case Some(childItem) =>
+                  saveNodeAndCreateItem(childNode.updated(childItemIdx, childItem), commPrefix).some
+              }
+            } yield returnedChild
+          }
+      }
+
+    /**
+      * Parallel processing of HistoryActions in this part of tree (start from curNode).
+      * New data load to [[cacheW]].
+      * Return updated curNode. if no action was taken - return [[None]].
+      */
+    def makeActions(curNode: Node, curActions: List[HistoryAction]): F[Option[Node]] = {
+      val groups = curActions
+        .groupBy( //fixme
+          _.key.headOption.getOrElse({
+            assert(assertion = false, "The length of all prefixes in the subtree must be the same")
+            0x00.toByte
+          })
+        )
+        .toList
+      val newGroupItemsF = groups.map {
+        case (groupIdx, groupActions) =>
+          val index = byteToInt(groupIdx)
+          val item  = curNode(index)
+          if (groupActions.length == 1) {
```

Comment:
Just a question. Why operation of `groupActions.length == 1` is different from `!=1`? I think they could be merged into same flow of operations.

Reply:
This is the key idea of parallel traverse algorithm:
![image](https://user-images.githubusercontent.com/81892279/147820618-c3122061-c1ea-41fd-af20-4f8092cef583.png)
If we have `==1` we can't parallel next and we should use sequential traversing with help insert() or delete().
If we have `!=1` we can create more parallel processes.